File: probability_model.py
import math
from game_constants import (
    ATTR_EFFECT_MIDPOINT,
    AVG_K_RATE_AT_MIDPOINT, MIN_K_RATE, MAX_K_RATE,
    K_RATE_EYE_WEIGHT, K_RATE_HIT_WEIGHT,
    K_EYE_EFFECT_MIDPOINT, K_EYE_EFFECT_SCALE, K_HIT_EFFECT_MIDPOINT, K_HIT_EFFECT_SCALE,
    AVG_BB_RATE_AT_MIDPOINT, MIN_BB_RATE, MAX_BB_RATE,
    BB_EYE_EFFECT_MIDPOINT, BB_EYE_EFFECT_SCALE,
    
    # New HR Model Parameters
    HR_S_CURVE_POW_ANCHORS, ABSOLUTE_MAX_HR_RATE_CAP,
    HR_EYE_MODIFIER_MIDPOINT, HR_EYE_MODIFIER_SCALE, HR_EYE_MODIFIER_MAX_IMPACT,
    HR_HIT_MODIFIER_MIDPOINT, HR_HIT_MODIFIER_SCALE, HR_HIT_MODIFIER_MAX_IMPACT,

    AVG_BABIP_AT_MIDPOINT, MIN_BABIP_RATE, MAX_BABIP_RATE,
    BABIP_HIT_EFFECT_MIDPOINT, BABIP_HIT_EFFECT_SCALE,
    
    AVG_2B_PER_HIT_BIP_NOT_HR_AT_MIDPOINT, MIN_2B_PER_HIT_BIP_NOT_HR, MAX_2B_PER_HIT_BIP_NOT_HR,
    EXTRABASE_POW_EFFECT_MIDPOINT, EXTRABASE_POW_EFFECT_SCALE,
    EXTRABASE_HIT_EFFECT_MIDPOINT, EXTRABASE_HIT_EFFECT_SCALE,
    EXTRABASE_POW_WEIGHT, EXTRABASE_HIT_WEIGHT,
    LEAGUE_AVG_HBP_RATE
)

def scale_attribute_to_effectiveness(attribute_value, midpoint, scale, effect_is_positive=True):
    """
    Scales an attribute to an effectiveness factor, typically between -1 and 1 (for tanh).
    'effect_is_positive': True if higher attribute means "more of the good thing / less of the bad thing".
    """
    if scale == 0:
        return 0.0
    normalized_value = (attribute_value - midpoint) / scale
    tanh_val = math.tanh(normalized_value)
    return tanh_val if effect_is_positive else -tanh_val

def get_rate_from_effectiveness(base_rate_at_midpoint, min_rate, max_rate, effectiveness_factor):
    """
    Calculates a rate based on an effectiveness factor (-1 to +1).
    """
    if effectiveness_factor >= 0:
        return base_rate_at_midpoint + effectiveness_factor * (max_rate - base_rate_at_midpoint)
    else:
        return base_rate_at_midpoint + effectiveness_factor * (base_rate_at_midpoint - min_rate)

def calculate_k_rate(EYE, HIT):
    """Calculates K rate based on EYE and HIT."""
    eye_eff_k = scale_attribute_to_effectiveness(EYE, K_EYE_EFFECT_MIDPOINT, K_EYE_EFFECT_SCALE, effect_is_positive=False) # Higher EYE = lower K%
    hit_eff_k = scale_attribute_to_effectiveness(HIT, K_HIT_EFFECT_MIDPOINT, K_HIT_EFFECT_SCALE, effect_is_positive=False) # Higher HIT = lower K% (better contact)
    
    # Weighted average of effectiveness factors
    # Note: effect_is_positive=False means tanh_val is already inverted for "bad" outcomes.
    # So, a positive tanh_val (good attribute) becomes negative here, pushing towards MIN_K_RATE.
    # A negative tanh_val (bad attribute) becomes positive, pushing towards MAX_K_RATE.
    # To align with get_rate_from_effectiveness, where positive effectiveness means towards max_rate (good thing)
    # and negative means towards min_rate (bad thing), we need to ensure the interpretation is consistent.
    # For K_Rate, MAX_K_RATE is "bad" and MIN_K_RATE is "good".
    # So, if eye_eff_k is positive (meaning EYE is high, good for reducing K), it should push towards MIN_K_RATE.
    # The current scale_attribute_to_effectiveness with effect_is_positive=False returns:
    #   - positive value if attribute is low (bad for K%) -> pushes towards MAX_K_RATE (correct)
    #   - negative value if attribute is high (good for K%) -> pushes towards MIN_K_RATE (correct)
    combined_eff_k = K_RATE_EYE_WEIGHT * eye_eff_k + K_RATE_HIT_WEIGHT * hit_eff_k
    
    k_rate = get_rate_from_effectiveness(AVG_K_RATE_AT_MIDPOINT, MIN_K_RATE, MAX_K_RATE, combined_eff_k)
    return max(MIN_K_RATE, min(MAX_K_RATE, k_rate))


def calculate_bb_rate(EYE):
    """Calculates BB rate based on EYE."""
    eye_eff_bb = scale_attribute_to_effectiveness(EYE, BB_EYE_EFFECT_MIDPOINT, BB_EYE_EFFECT_SCALE, effect_is_positive=True)
    bb_rate = get_rate_from_effectiveness(AVG_BB_RATE_AT_MIDPOINT, MIN_BB_RATE, MAX_BB_RATE, eye_eff_bb)
    return max(MIN_BB_RATE, min(MAX_BB_RATE, bb_rate))

def interpolate_s_curve(value, anchors):
    """
    Performs linear interpolation for an S-curve defined by anchor points.
    Anchors should be a list of (x, y) tuples, sorted by x.
    """
    if not anchors:
        return 0.0
    
    # Handle cases where value is outside the range of anchors
    if value <= anchors[0][0]:
        return anchors[0][1]
    if value >= anchors[-1][0]:
        return anchors[-1][1]

    # Find the two anchor points to interpolate between
    for i in range(len(anchors) - 1):
        x1, y1 = anchors[i]
        x2, y2 = anchors[i+1]
        if x1 <= value < x2:
            # Linear interpolation formula: y = y1 + (y2 - y1) * (value - x1) / (x2 - x1)
            if (x2 - x1) == 0: # Avoid division by zero, should not happen if anchors are distinct
                return y1 
            return y1 + (y2 - y1) * (value - x1) / (x2 - x1)
    
    return anchors[-1][1] # Should be covered by boundary checks, but as a fallback

def calculate_base_hr_rate_from_pow_s_curve(POW):
    """
    Calculates the base HR rate (per PA) using an S-curve based on POW.
    Uses linear interpolation between anchor points defined in game_constants.
    """
    base_hr_rate = interpolate_s_curve(POW, HR_S_CURVE_POW_ANCHORS)
    return base_hr_rate

def get_pa_event_probabilities(POW, HIT, EYE, player_hbp_rate):
    """
    Calculates the probabilities of different plate appearance outcomes
    using the new sequential model for HR.
    """
    # 1. Calculate K%, BB%, HBP%
    p_k = calculate_k_rate(EYE, HIT)
    p_bb = calculate_bb_rate(EYE)
    
    # Ensure player_hbp_rate is within reasonable bounds, e.g., 0 to 0.05
    p_hbp = max(0.0, min(0.05, player_hbp_rate if player_hbp_rate is not None else LEAGUE_AVG_HBP_RATE))

    # 2. Calculate new P(HR) based on S-curve(POW) and EYE/HIT modifiers
    base_p_hr_pa = calculate_base_hr_rate_from_pow_s_curve(POW)

    # EYE Modifier for HR
    eye_eff_hr_mod = scale_attribute_to_effectiveness(EYE, HR_EYE_MODIFIER_MIDPOINT, HR_EYE_MODIFIER_SCALE, True)
    eye_modifier = 1.0 + (eye_eff_hr_mod * HR_EYE_MODIFIER_MAX_IMPACT) # Centered around 1.0

    # HIT Modifier for HR
    hit_eff_hr_mod = scale_attribute_to_effectiveness(HIT, HR_HIT_MODIFIER_MIDPOINT, HR_HIT_MODIFIER_SCALE, True)
    hit_modifier = 1.0 + (hit_eff_hr_mod * HR_HIT_MODIFIER_MAX_IMPACT) # Centered around 1.0

    p_hr = base_p_hr_pa * eye_modifier * hit_modifier
    p_hr = max(0.0, min(p_hr, ABSOLUTE_MAX_HR_RATE_CAP)) # Apply absolute cap

    # 3. Calculate remaining probability for other BIP outcomes (1B, 2B, IPO)
    prob_sum_non_bip_plus_hr = p_k + p_bb + p_hbp + p_hr
    
    # Safety clamp: if these events already sum to 1 or more, adjust.
    if prob_sum_non_bip_plus_hr >= 1.0:
        # Need to re-normalize K, BB, HBP, HR if they exceed 1.0
        # For simplicity here, we'll cap them and assign 0 to BIP outcomes.
        # A more robust solution would scale them down proportionally.
        scale_down = 1.0 / prob_sum_non_bip_plus_hr if prob_sum_non_bip_plus_hr > 0 else 1.0
        p_k *= scale_down
        p_bb *= scale_down
        p_hbp *= scale_down
        p_hr *= scale_down
        p_hr = max(0, p_hr) # re-ensure after scaling
        
        p_bip_for_other_outcomes = 0.0
        p_1b = 0.0
        p_2b = 0.0
        p_ipo = 0.0
    else:
        p_bip_for_other_outcomes = 1.0 - prob_sum_non_bip_plus_hr

        # 4. Calculate BABIP for the remaining BIP events
        hit_eff_babip = scale_attribute_to_effectiveness(HIT, BABIP_HIT_EFFECT_MIDPOINT, BABIP_HIT_EFFECT_SCALE, True)
        # p_hit_given_bip_remaining is P(Hit | BIP_for_other_outcomes)
        p_hit_given_bip_remaining = get_rate_from_effectiveness(AVG_BABIP_AT_MIDPOINT, MIN_BABIP_RATE, MAX_BABIP_RATE, hit_eff_babip)
        p_hit_given_bip_remaining = max(MIN_BABIP_RATE, min(MAX_BABIP_RATE, p_hit_given_bip_remaining))

        # 5. Calculate P(Total Hits on these BIPs) and P(IPO on these BIPs)
        p_total_hits_on_remaining_bip = p_bip_for_other_outcomes * p_hit_given_bip_remaining
        p_ipo = p_bip_for_other_outcomes * (1.0 - p_hit_given_bip_remaining)
        p_ipo = max(0, p_ipo)

        # 6. Distribute p_total_hits_on_remaining_bip into P(1B) and P(2B)
        # (Triples are not explicitly modeled, implicitly part of 1B or 2B)
        if p_total_hits_on_remaining_bip > 0:
            pow_eff_xbh = scale_attribute_to_effectiveness(POW, EXTRABASE_POW_EFFECT_MIDPOINT, EXTRABASE_POW_EFFECT_SCALE, True)
            hit_eff_xbh = scale_attribute_to_effectiveness(HIT, EXTRABASE_HIT_EFFECT_MIDPOINT, EXTRABASE_HIT_EFFECT_SCALE, True)
            combined_eff_xbh = EXTRABASE_POW_WEIGHT * pow_eff_xbh + EXTRABASE_HIT_WEIGHT * hit_eff_xbh
            
            # p_2b_given_hit_bip_not_hr is P(2B | Hit_on_remaining_BIP)
            p_2b_given_hit_bip_not_hr = get_rate_from_effectiveness(
                AVG_2B_PER_HIT_BIP_NOT_HR_AT_MIDPOINT, 
                MIN_2B_PER_HIT_BIP_NOT_HR, 
                MAX_2B_PER_HIT_BIP_NOT_HR, 
                combined_eff_xbh
            )
            p_2b_given_hit_bip_not_hr = max(MIN_2B_PER_HIT_BIP_NOT_HR, min(MAX_2B_PER_HIT_BIP_NOT_HR, p_2b_given_hit_bip_not_hr))
            
            p_2b = p_total_hits_on_remaining_bip * p_2b_given_hit_bip_not_hr
            p_1b = p_total_hits_on_remaining_bip * (1.0 - p_2b_given_hit_bip_not_hr)
            p_2b = max(0, p_2b)
            p_1b = max(0, p_1b)
        else:
            p_1b = 0.0
            p_2b = 0.0
            
    # Final check for negative probabilities due to floating point arithmetic
    p_hr = max(0.0, p_hr)
    p_2b = max(0.0, p_2b)
    p_1b = max(0.0, p_1b)
    p_bb = max(0.0, p_bb)
    p_hbp = max(0.0, p_hbp)
    p_k = max(0.0, p_k)
    p_ipo = max(0.0, p_ipo)

    # 7. Normalize all probabilities to sum to 1.0
    events = {
        "HR": p_hr, "2B": p_2b, "1B": p_1b,
        "BB": p_bb, "HBP": p_hbp, "K": p_k, "IPO": p_ipo
    }
    
    current_total_prob = sum(events.values())
    
    if current_total_prob == 0: # Should be extremely rare
        # Fallback to a safe state, e.g., 100% IPO if all probabilities somehow became zero.
        return {"HR": 0, "2B": 0, "1B": 0, "BB": 0, "HBP": 0, "K": 0, "IPO": 1.0}

    # Normalize
    norm_factor = 1.0 / current_total_prob
    normalized_events = {key: value * norm_factor for key, value in events.items()}
    
    # Final check of sum after normalization (for debugging)
    # final_sum_check = sum(normalized_events.values())
    # if abs(final_sum_check - 1.0) > 1e-5:
    #     print(f"Warning: Final normalization sum is {final_sum_check}")

    return normalized_events

File: probability_model.js
// probability-model.js - 概率計算模型（強化極端值版）

// 輔助函數：S-curve 插值
function interpolateSCurve(value, anchors) {
    if (!anchors || !anchors.length) return 0.0;
    if (value <= anchors[0][0]) return anchors[0][1];
    if (value >= anchors[anchors.length - 1][0]) return anchors[anchors.length - 1][1];
    
    for (let i = 0; i < anchors.length - 1; i++) {
        const [x1, y1] = anchors[i];
        const [x2, y2] = anchors[i + 1];
        if (x1 <= value && value < x2) {
            return (x2 - x1) ? y1 + (y2 - y1) * (value - x1) / (x2 - x1) : y1;
        }
    }
    return anchors[anchors.length - 1][1];
}

// 輔助函數：屬性值轉換為效果因子
function scaleAttributeToEffectiveness(attributeValue, midpoint, scale, effectIsPositive = true) {
    if (scale === 0) return 0.0;
    const normalizedValue = (attributeValue - midpoint) / scale;
    const tanhVal = Math.tanh(normalizedValue);
    return effectIsPositive ? tanhVal : -tanhVal;
}

// 輔助函數：從效果因子計算比率
function getRateFromEffectiveness(baseRateAtMidpoint, minRate, maxRate, effectivenessFactor) {
    return effectivenessFactor >= 0 
        ? baseRateAtMidpoint + effectivenessFactor * (maxRate - baseRateAtMidpoint)
        : baseRateAtMidpoint + effectivenessFactor * (baseRateAtMidpoint - minRate);
}

// 🔥 正常範圍的概率計算（完全保持原版邏輯）
function getPAEventProbabilitiesNormal(POW, HIT, EYE, playerHBPRate = LEAGUE_AVG_HBP_RATE) {
    // 計算 K%, BB%, HBP%
    const eyeKEffect = interpolateSCurve(EYE, K_EYE_EFFECTIVENESS_S_CURVE_ANCHORS);
    const hitKEffect = scaleAttributeToEffectiveness(HIT, K_HIT_EFFECT_MIDPOINT, K_HIT_EFFECT_SCALE, false);
    const combinedKEffect = K_RATE_EYE_WEIGHT * eyeKEffect + K_RATE_HIT_WEIGHT * hitKEffect;
    const pK = Math.max(MIN_K_RATE_CAP, Math.min(MAX_K_RATE_CAP, 
        getRateFromEffectiveness(AVG_K_RATE_AT_MIDPOINT, MIN_K_RATE_CAP, MAX_K_RATE_CAP, combinedKEffect)));
    
    const pBB = Math.max(0.020, Math.min(0.250, interpolateSCurve(EYE, BB_S_CURVE_EYE_ANCHORS)));
    const pHBP = Math.max(0.0, Math.min(0.05, playerHBPRate));
    
    // 計算基礎全壘打率
    let baseHRRate = interpolateSCurve(POW, HR_S_CURVE_POW_ANCHORS);
    
    // 應用 EYE 和 HIT 修正因子（最小化疊加效應以精確匹配分析表）
    const eyeHRModifier = 1.0 + scaleAttributeToEffectiveness(EYE, 70.0, 40.0, true) * 0.02; // 進一步降低: 0.04→0.02
    const hitHRModifier = 1.0 + scaleAttributeToEffectiveness(HIT, 70.0, 40.0, true) * 0.03; // 進一步降低: 0.06→0.03
    let pHR = baseHRRate * eyeHRModifier * hitHRModifier;
    pHR = Math.max(0.0, Math.min(pHR, 0.20));
    
    // 計算剩餘 BIP 事件
    const probSumNonBIPPlusHR = pK + pBB + pHBP + pHR;
    
    let p1B, p2B, pIPO;
    if (probSumNonBIPPlusHR >= 1.0) {
        const scaleDown = 1.0 / probSumNonBIPPlusHR;
        pHR = Math.max(0, 1.0 - (pK * scaleDown + pBB * scaleDown + pHBP * scaleDown));
        p1B = p2B = pIPO = 0.0;
    } else {
        const pBIPForOtherOutcomes = 1.0 - probSumNonBIPPlusHR;
        const pHitGivenBIPRemaining = Math.max(0.190, Math.min(0.450, 
            interpolateSCurve(HIT, BABIP_S_CURVE_HIT_ANCHORS)));
        const pTotalHitsOnRemainingBIP = pBIPForOtherOutcomes * pHitGivenBIPRemaining;
        pIPO = Math.max(0, pBIPForOtherOutcomes * (1.0 - pHitGivenBIPRemaining));
        
        if (pTotalHitsOnRemainingBIP > 0) {
            // 🔥 使用 POW-dependent 2B rate system
            const p2BGivenHitBIPNotHR = interpolateSCurve(POW, DOUBLES_RATE_S_CURVE_POW_ANCHORS);
            const p2BGivenHitBIPNotHRCapped = Math.max(MIN_2B_PER_HIT_BIP_NOT_HR, 
                Math.min(MAX_2B_PER_HIT_BIP_NOT_HR, p2BGivenHitBIPNotHR));
            
            p2B = Math.max(0, pTotalHitsOnRemainingBIP * p2BGivenHitBIPNotHRCapped);
            p1B = Math.max(0, pTotalHitsOnRemainingBIP * (1.0 - p2BGivenHitBIPNotHRCapped));
        } else {
            p1B = p2B = 0.0;
        }
    }
    
    return {HR: pHR, '2B': p2B, '1B': p1B, BB: pBB, HBP: pHBP, K: pK, IPO: pIPO};
}

// 🔥 極端值的概率計算（大幅強化）
function getPAEventProbabilitiesExtreme(POW, HIT, EYE, playerHBPRate = LEAGUE_AVG_HBP_RATE) {
    console.log(`🔥 極端值計算: POW=${POW}, HIT=${HIT}, EYE=${EYE}`);
    
    // 🔥 強化：使用更激進的極端值 S-curves
    let pHR = interpolateSCurve(POW, HR_S_CURVE_POW_ANCHORS_EXTREME);
    let babip = interpolateSCurve(HIT, BABIP_S_CURVE_HIT_ANCHORS_EXTREME);
    let pBB = interpolateSCurve(EYE, BB_S_CURVE_EYE_ANCHORS_EXTREME);
    
    // 🔥 強化：極端值時的額外加成
    if (POW >= 400) {
        pHR = Math.min(0.99, pHR * 1.2); // 400+ POW 時額外 20% 加成
    }
    if (HIT >= 400) {
        babip = Math.min(0.995, babip * 1.1); // 400+ HIT 時額外 10% 加成
    }
    if (EYE >= 400) {
        pBB = Math.min(0.98, pBB * 1.1); // 400+ EYE 時額外 10% 加成
    }
    
    // 三振率計算 - 極端值時大幅改善
    const eyeKEffect = interpolateSCurve(EYE, K_EYE_EFFECTIVENESS_S_CURVE_ANCHORS_EXTREME);
    const hitKEffect = scaleAttributeToEffectiveness(HIT, K_HIT_EFFECT_MIDPOINT, 55.0, false);
    let kRate = getRateFromEffectiveness(
        AVG_K_RATE_AT_MIDPOINT, MIN_K_RATE_CAP, MAX_K_RATE_CAP,
        K_RATE_EYE_WEIGHT * eyeKEffect + K_RATE_HIT_WEIGHT * hitKEffect
    );
    
    // 🔥 強化：極端值時三振率幾乎歸零
    if (HIT >= 300 || EYE >= 300) {
        kRate = Math.max(0.001, kRate * 0.05); // 只有 5% 的原三振率
    }
    if (HIT >= 450 && EYE >= 450) {
        kRate = 0.001; // 幾乎不三振
    }
    
    // HBP 極端值時稍微提高
    let pHBP = LEAGUE_AVG_HBP_RATE;
    if (EYE >= 350) {
        pHBP = Math.min(0.050, LEAGUE_AVG_HBP_RATE * 3);
    }
    
    // 🔥 強化：確保極端值時能達到理論極限
    const basicSum = pHR + pBB + pHBP + kRate;
    
    if (basicSum >= 1.0) {
        // 極端值時優先保證核心表現
        const totalAvailable = 0.999; // 保留一點給其他事件
        
        // 按重要性分配概率
        if (POW >= 450) pHR = Math.min(0.97, pHR); // 97% 全壘打率
        if (EYE >= 450) pBB = Math.min(0.95, pBB);  // 95% 保送率
        
        // 重新分配
        const newSum = pHR + pBB + pHBP + kRate;
        if (newSum > totalAvailable) {
            const scale = totalAvailable / newSum;
            pHR *= scale;
            pBB *= scale;
            kRate *= scale;
        }
        
        const remainingProb = 1.0 - pHR - pBB - pHBP - kRate;
        
        return {
            HR: pHR,
            '2B': remainingProb * 0.8, // 極端值時大部分是長打
            '1B': remainingProb * 0.2,
            BB: pBB,
            HBP: pHBP,
            K: kRate,
            IPO: 0.001
        };
    }
    
    // 計算剩餘 BIP 事件
    const remainingBIP = 1.0 - pHR - pBB - pHBP - kRate;
    const pHitFromBIP = remainingBIP * babip;
    const pIPO = remainingBIP * (1.0 - babip);
    
    // 🔥 強化：極端值時大幅提高長打比例
    let extrabaseRatio = 0.3; // 基準比例
    if (POW >= 300) extrabaseRatio = Math.min(0.85, 0.3 + (POW - 300) / 400);
    if (HIT >= 300) extrabaseRatio = Math.min(0.90, extrabaseRatio + (HIT - 300) / 500);
    if (POW >= 450) extrabaseRatio = 0.95; // 450+ POW 時 95% 是長打
    
    const p2B = pHitFromBIP * extrabaseRatio;
    const p1B = pHitFromBIP * (1 - extrabaseRatio);
    
    const result = {
        HR: pHR,
        '2B': p2B,
        '1B': p1B,
        BB: pBB,
        HBP: pHBP,
        K: kRate,
        IPO: pIPO
    };
    
    console.log(`🔥 強化極端值結果:`, result);
    return result;
}

// 🔥 主要的概率計算函數（智能判斷使用哪種計算方式）
function getPAEventProbabilities(POW, HIT, EYE, playerHBPRate = LEAGUE_AVG_HBP_RATE) {
    // 只有當任一屬性值 >= 200 時才使用極端值計算
    const isExtreme = POW >= EXTREME_VALUE_THRESHOLD || HIT >= EXTREME_VALUE_THRESHOLD || EYE >= EXTREME_VALUE_THRESHOLD;
    
    if (isExtreme) {
        console.log(`🔥 檢測到極端屬性，使用強化極端值計算: POW=${POW}, HIT=${HIT}, EYE=${EYE}`);
        return getPAEventProbabilitiesExtreme(POW, HIT, EYE, playerHBPRate);
    }
    
    // 正常值使用原版精確計算
    return getPAEventProbabilitiesNormal(POW, HIT, EYE, playerHBPRate);
}